# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [ ]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf
import statsmodels.api as sm

tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [ ]:
model_padrao = smf.ols('tip_pct ~ net_bill', data=tips).fit()
print(model_padrao.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.091
Model:                            OLS   Adj. R-squared:                  0.088
Method:                 Least Squares   F-statistic:                     24.31
Date:                Tue, 13 Aug 2024   Prob (F-statistic):           1.52e-06
Time:                        17:59:00   Log-Likelihood:                 108.00
No. Observations:                 244   AIC:                            -212.0
Df Residuals:                     242   BIC:                            -205.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3053      0.023     13.166      0.0

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [ ]:
model = smf.ols('tip_pct ~ np.log(net_bill)', data=tips).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Tue, 13 Aug 2024   Prob (F-statistic):           1.62e-13
Time:                        17:49:32   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.6181      0.054  

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [ ]:
tips['net_bill_sq'] = tips['net_bill'] ** 2

# Ajustando o modelo de tip_pct em função de net_bill e net_bill_sq
model_polinomio = smf.ols('tip_pct ~ net_bill + net_bill_sq', data=tips).fit()

# Resumo do modelo
print(model_polinomio.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Tue, 13 Aug 2024   Prob (F-statistic):           1.06e-09
Time:                        17:50:34   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept       0.4851      0.047     10.332      

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [ ]:
model_tip_pct_log = smf.ols('np.log(tip_pct) ~ net_bill', data=tips).fit()

print(model_tip_pct_log.summary())

                            OLS Regression Results                            
Dep. Variable:        np.log(tip_pct)   R-squared:                       0.235
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     74.48
Date:                Tue, 13 Aug 2024   Prob (F-statistic):           8.27e-16
Time:                        17:52:14   Log-Likelihood:                -106.00
No. Observations:                 244   AIC:                             216.0
Df Residuals:                     242   BIC:                             223.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.2749      0.056    -22.872      0.0

In [ ]:
tips['predicted_tip_pct'] = np.exp(model_tip_pct_log.predict(tips['net_bill']))

# Calculando o R² para tip_pct original
SS_res = np.sum((tips['tip_pct'] - tips['predicted_tip_pct']) ** 2)
SS_tot = np.sum((tips['tip_pct'] - np.mean(tips['tip_pct'])) ** 2)
r_squared = 1 - (SS_res / SS_tot)

print(f"R² para tip_pct original: {r_squared:.4f}")

R² para tip_pct original: 0.0868


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [ ]:
model_net_bill_log = smf.ols('np.log(tip_pct) ~ np.log(net_bill)', data=tips).fit()

# Resumo do modelo
print(model_net_bill_log.summary())


                            OLS Regression Results                            
Dep. Variable:        np.log(tip_pct)   R-squared:                       0.293
Model:                            OLS   Adj. R-squared:                  0.291
Method:                 Least Squares   F-statistic:                     100.5
Date:                Tue, 13 Aug 2024   Prob (F-statistic):           5.20e-20
Time:                        17:54:07   Log-Likelihood:                -96.357
No. Observations:                 244   AIC:                             196.7
Df Residuals:                     242   BIC:                             203.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           -0.3942      0.133  

In [ ]:
# Previsão de tip_pct usando o modelo ajustado (em escala original)
tips['predicted_tip_pct'] = np.exp(model_net_bill_log.predict(np.log(tips['net_bill'])))

# Calculando o R² para tip_pct original
SS_res = np.sum((tips['tip_pct'] - tips['predicted_tip_pct']) ** 2)
SS_tot = np.sum((tips['tip_pct'] - np.mean(tips['tip_pct'])) ** 2)
r_squared = 1 - (SS_res / SS_tot)

print(f"R² para tip_pct original: {r_squared:.4f}")

R² para tip_pct original: -1.5288


###  6. qual modelo resultou no melhor ajuste?

R² nas diversas variações.

* **R quadrado inicial:** 0.091
* **R quadrado *log de net_bill*:** 0.202
* **R quadrado *polinômio de 2º em net_bill*:** 0.158
* **R quadrado *log tip_pct*:** 0.235
* **R quadrado *log tip_pct e net_bill*:** 0.293

O que apresentou o melhor ajuste foi utilizar o *log* em ambas as variaveis ( tip_pct e net_bill )
